pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/OncHistoryUploadRoute.java

```diff
@@ -0,0 +1,75 @@
+package org.broadinstitute.dsm.route;
+
+import lombok.extern.slf4j.Slf4j;
+import org.broadinstitute.dsm.db.dao.user.UserDao;
+import org.broadinstitute.dsm.db.dto.user.UserDto;
+import org.broadinstitute.dsm.exception.DSMBadRequestException;
+import org.broadinstitute.dsm.exception.DsmInternalError;
+import org.broadinstitute.dsm.security.RequestHandler;
+import org.broadinstitute.dsm.service.onchistory.CodeStudyColumnsProvider;
+import org.broadinstitute.dsm.service.onchistory.OncHistoryUploadService;
+import org.broadinstitute.dsm.service.onchistory.OncHistoryValidationException;
+import org.broadinstitute.dsm.statics.RoutePath;
+import org.broadinstitute.dsm.statics.UserErrorMessages;
+import org.broadinstitute.dsm.util.UserUtil;
+import spark.QueryParamsMap;
+import spark.Request;
+import spark.Response;
+
+@Slf4j
+public class OncHistoryUploadRoute extends RequestHandler {
+
+    public static final String ONC_HISTORY_UPLOAD = "kit_upload"; // TOOO temp until testers are part of role -DC
+    //public static final String ONC_HISTORY_UPLOAD = "onc_history_upload";
+
+    @Override
+    protected Object processRequest(Request request, Response response, String userId) throws Exception {
+        QueryParamsMap queryParams = request.queryMap();
+        String realm;
+        if (!queryParams.hasKey(RoutePath.REALM)) {
+            response.status(400);
+            return "Request requires realm parameter";
+        }
+        realm = queryParams.value(RoutePath.REALM);
+
+        if (!canUploadOncHistory(realm, userId)) {
+            response.status(401);
+            return (UserErrorMessages.NO_RIGHTS);
+        }
+
+        String oncHistoryUserId;
+        try {
+            UserDto user = new UserDao().get(Integer.parseInt(userId)).orElseThrow();
+            oncHistoryUserId = user.getEmail().orElseThrow();
+            if (oncHistoryUserId.isEmpty()) {
+                throw new DsmInternalError("Empty email address");
+            }
+        } catch (Exception e) {
+            response.status(500);
+            return "No email address for user " + userId;
+        }
+
+        try {
+            OncHistoryUploadService service =
+                    new OncHistoryUploadService(realm, oncHistoryUserId, new CodeStudyColumnsProvider());
+            service.upload(request.body());
+            response.status(200);
+            // TODO: needed?
+            return "Onc history upload succeeded";
+        } catch (DSMBadRequestException | OncHistoryValidationException e) {
```

Comment:
Can't these be set up in DSMServer per the docs here?
http://sparkjava.com/documentation.html#exception-mapping

This way instead of a particular route having to trap the error, the Spark framework can catch from all routes.

Reply:
Yes, they can.
I want to do a broader cleanup of our exception to response status code mapping and will include this case then.
Do you think a change at this time is warranted?